routers/web/admin/pprof.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package admin
+
+import (
+	"fmt"
+	"net/http"
+	"runtime/pprof"
+	"runtime/trace"
+	"strconv"
+	"time"
+
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/process"
+	"code.gitea.io/gitea/modules/setting"
+
+	"github.com/felixge/fgprof"
+)
+
+// PProfProcessStacktrace returns the stacktrace similar to GoroutineStacktrace but without rendering it
+func PProfProcessStacktrace(ctx *context.Context) {
+	flat := ctx.FormBool("flat")
+	noSystem := ctx.FormBool("no-system")
+
+	format := ctx.FormString("format")
+	jsonFormat := format == "json"
+
+	start := time.Now()
+	filename := "process-stacktrace-" + strconv.FormatInt(start.Unix(), 10)
+	if jsonFormat {
+		filename += ".json"
+	}
+
+	processStacks, processCount, goroutineCount, err := process.GetManager().ProcessStacktraces(flat, noSystem)
+	if err != nil {
+		ctx.ServerError("ProcessStacktraces", err)
+	}
+
+	ctx.SetServeHeaders(&context.ServeHeaderOptions{
+		Filename:     filename,
+		LastModified: start,
+	})
+
+	if jsonFormat {
+		ctx.JSON(http.StatusOK, map[string]interface{}{
+			"TotalNumberOfGoroutines": goroutineCount,
+			"TotalNumberOfProcesses":  processCount,
+			"Processes":               processStacks,
+		})
+		return
+	}
+
+	if err := process.WriteProcesses(ctx.Resp, processStacks, processCount, goroutineCount, "", flat); err != nil {
+		ctx.ServerError("WriteProcesses", err)
+		return
+	}
+}
+
+// PProfFGProfile returns the Full Go Profile from fgprof
+func PProfFGProfile(ctx *context.Context) {
+	durationStr := ctx.FormString("duration")
+	duration := 30 * time.Second
+	if durationStr != "" {
+		var err error
+		duration, err = time.ParseDuration(durationStr)
+		if err != nil {
+			ctx.Flash.Error(ctx.Tr("admin.monitor.pprof.duration_invalid"))
+			ctx.Redirect(setting.AppSubURL + "/admin/monitor")
+			return
+		}
+	}
+
+	format := fgprof.Format(ctx.FormString("format"))
+	if format != fgprof.FormatFolded {
+		format = fgprof.FormatPprof
+	}
```

Comment:
I have some difficulty understanding this logic. Why?
at best add a comment about it.

Reply:
What's difficult to understand here? If you provide folded then it's the folded profile otherwise it's the pprof profile

Reply:
Yeah, but why would you reassign it? Are there other possibilities we ignore?
Is the `fgprof.FormatPprof` not assigned through the function above?
Or why would you manually need to reassign it?

Reply:
There are no other possibilities. There is only FormatFolded or FormatPprof and FormatPprof is the default. So if the provided format isn't folded set it to Pprof.

I don't understand the problem - that is the only interpretation of that code that is possible.

Reply:
Ah, I think I understand.
`Format()` is simply a type cast.
To me, it looked like a function that would automatically return the correct type.
That's why I was confused.
But with a type cast, it of course makes sense.